src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,236 @@
+            node = node.next;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
+        return size == 0;
+    }
+
+    private T removeLast() {
+        Node<T> nodeToRemove = endNode;
+        unlink(endNode);
+        return nodeToRemove.data;
+    }
+
+    private T removeFirst() {
+        Node<T> nodeToRemove = startNode;
+        unlink(startNode);
+        return nodeToRemove.data;
+    }
+
+    private void unlink(Node<T> node) {
+        if (node.previous == null) {
+            startNode = startNode.next;
+            size--;
+        } else {
+            endNode = endNode.previous;
+            size--;
+        }
+    }
+
+    private T removeByIndex(int index) {
+        Node<T> node = findByIndex(index);
+        if (size == 2) {
+            if (index == 1) {
+                startNode.next = null;
+                endNode = null;
+            } else if (index == 0) {
+                startNode = endNode;
+                endNode = null;
+                startNode.next = null;
+            }
+        } else if (size == 1) {
+            node = startNode;
+            startNode = null;
+        } else {
+            node.previous.next = node.next;
+            node.next.previous = node.previous;
+        }
+        size--;
+        return node.data;
+    }
+
+    private void linkLast(Node<T> node) {
+        node.previous = endNode;
+        endNode.next = node;
+        endNode = node;
+        size++;
+    }
+
+    private void linkFirst(Node<T> node) {
+        startNode.previous = node;
+        node.next = startNode;
+        startNode = node;
+        size++;
+    }
+
+    private void linkAtMiddle(Node<T> newNode, int index) {
+        Node<T> existingNode = findByIndex(index);
+        newNode.previous = existingNode.previous;
+        existingNode.previous.next = newNode;
+        newNode.next = existingNode;
+        existingNode.previous = newNode;
+        size++;
+    }
+
+    private Node<T> findByIndex(int index) {
+        indexValidation(index);
+        if (index >= 0 && index < 2 && size < 2) {
+            return index == 0 ? startNode : endNode;
+        }
+        if (index - size / 2 <= 0) {
+            return findNodeFromStart(index);
+        } else {
+            return findNodeFromEnd(index);
+        }
+    }
+
+    private Node<T> findNodeFromStart(int index) {
+        Node<T> x = startNode;
+        int currentIndex = 0;
+        while (index != currentIndex) {
+            x = x.next;
+            currentIndex++;
+        }
+        return x;
+    }
+
+    private Node<T> findNodeFromEnd(int index) {
+        Node<T> x = endNode;
+        int currentIndex = size - 1;
+        while (index != currentIndex) {
+            x = x.previous;
+            currentIndex--;
+        }
+        return x;
+    }
+
+    private void initializeBase(Node<T> node) {
```

Comment:
Minor. Just for clarity as you defined startNode and endNode

```suggestion
    private void initializeStartNode(Node<T> node) {
```